skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +174,341 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+
+        boundary = _boundary_case(u_total_count, v_total_count,
+                                  unweighted=False)
+        if boundary is not None:
+            return boundary
+
+        u = _weighted_unifrac(u_node_counts, v_node_counts, u_total_count,
+                              v_total_count, branch_lengths)
+        if normalized:
+            u /= _weighted_unifrac_branch_correction(
+                    node_to_root_distances, u_node_counts, v_node_counts,
+                    u_total_count, v_total_count)
+        return u
+
+    return f, counts_by_node, branch_lengths
+
+
+def _boundary_case(u_total_count, v_total_count, normalized=False,
+                   unweighted=True):
+    """Test for boundary conditions
+
+    Parameters
+    ----------
+    u_total_count, v_total_count : float
+        The sum of the observations in samples ``u`` and ``v``, respectively.
+    normalized : bool, optional
+        Indicates if the method is normalized.
+    unweighted : bool, optional
+        Indicates if the method is unweighted.
+
+    Returns
+    -------
+    float or None
+        Specifically, one of `[0.0, 1.0, None]`. `None` indicates that a
+        boundary condition was not observed, and the float values are distances
+        between ``u`` and ``v``.
+
+    Notes
+    -----
+    The following boundary conditions are tested:
+
+        * if u_total_count or v_total_count are zero
+        * if both u_total_count and v_total_count are zero
+    """
+    if u_total_count and v_total_count:
+        return None
+
+    if u_total_count + v_total_count:
+        # u or v counts are all zeros
+        if unweighted or normalized:
+            # u or v counts are all zeros
+            return 1.0
+        # NOTE: we cannot handle the unnormalized case here yet as it
+        # requires operations on the tree vector
     else:
-        return observed_nodes / total_count
+        # u and v are zero
+        return 0.0
+
+    return None
+
+
+def _weighted_unifrac_branch_correction(node_to_root_distances, u_node_counts,
+                                        v_node_counts, u_total_count,
+                                        v_total_count):
+    """Calculates weighted unifrac branch length correction.
+
+    Parameters
+    ----------
+    node_to_root_distances : np.ndarray
+        1D column vector of branch lengths in post order form. There should be
+        positions in this vector for all nodes in the tree, but only tips
+        should be non-zero.
+    u_node_counts, v_node_counts : np.ndarray
+        Counts of observations of all nodes in the tree in samples ``u`` and
+        ``v``, respectively.
+    u_total_count, v_total_count : float
+        The sum of the observations in samples ``u`` and ``v``, respectively.
+
+    Returns
+    -------
+    np.ndarray
+        The corrected branch lengths
+    """
+    return (node_to_root_distances.ravel() *
+            ((u_node_counts / u_total_count) +
```

Comment:
These should be returned from ``_weighted_unifrac`` so we don't have to recompute relative abundances here.